ai/worker: Handle LOADING state on containers healthcheck

victorges opened the pull request:
**What does this pull request do? Explain your changes. (required)**
This implements handling of the LOADING state on the worker code
and optimizes handling based on latest updates on https://github.com/livepeer/ai-runner/pull/568

By borrowing containers from the watch loop, we make sure the O will not
accept any job for that container while it is still loading.

**Specific updates (required)**
- Add LOADING state handling
- Refactor watch container logic a bit to react faster and be more lenient to new states we wanna create in the future

**How did you test each of these updates (required)**
box

**Does this pull request close any open issues?**
Included in INF-119

**Checklist:**
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] Read the [contribution guide](./CONTRIBUTING.md)
- [x] `make` runs successfully
- [x] All tests in `./test.sh` pass
- [ ] README and other documentation updated
- [ ] [Pending changelog](./CHANGELOG_PENDING.md) updated

victorges committed:
ai/worker: Be more lenient to unknown container statuses

We never know when we'll need to create another one, so let
them pass and just log.

victorges committed:
ai/worker: Add simple handling of LOADING state

victorges committed:
ai/worker: Fix startup timeouts for new LOADING state

victorges committed:
ai/worker: Make ERROR state restart immediately

We were waiting for 2 failures, but the ERROR state is
terminal and should be reacted immediately.

victorges committed:
TODO: reminder for another improvement

leszko commented on ai/worker/docker.go:
```diff
@@ -601,12 +605,30 @@ func (m *DockerManager) watchContainer(rc *RunnerContainer) {
 				failures = 0
 				continue
+			case "LOADING":
+				// TODO: Consider starting live containers as busy since they might take a while until the first healthcheck returns
```
What is this TODO, aren't you actually treating them as "busy" in this PR?

victorges replied:
Just implemented it, this is what I meant: https://github.com/livepeer/go-livepeer/pull/3569/commits/b43e63a28f5b035b7b8534d738a7bed83061a09e

Right now we do treat it as busy, but before there could be a slight race condition between starting the container as "idle" and then only on the first healthcheck from this function that we would move it to "busy" (borrowed). With the fix above, I made sure we don't even consider the container as created if it doesn't leave the LOADING state, which is done on a separate healthcheck logic (same API tho) done on `container.go` on container creation.

The logic here is still useful though, since we want to treat containers as busy if they end up restarting the pipeline process (which moves it back to [`LOADING` state here](https://github.com/livepeer/ai-runner/pull/568/files#diff-f1b7b77896c2b36646684e551871ccb2fced9456b2d7a7cbba8264c11411c22bR235)).

victorges replied:
Actually went on a different path for this to improve O startup time, otherwise we would wait for 1 container at a time to finish loading all the way until starting the rest of the containers and the rest of the O.

This is the change: https://github.com/livepeer/go-livepeer/pull/3569/commits/efeb744b7804ba82cadef4713d244bdafb3f221b

Also works fine (added a dummy sleep on `noop` startup):
```
2025/05/15 19:27:03 INFO Starting managed container gpu=colocated-0-emulated-0 name=live-video-to-video_noop_8900 modelID=noop containerImage=livepeer/ai-runner:live-app-noop
2025/05/15 19:27:11 INFO Warm container started on loading state, removing from pool on startup container=live-video-to-video_noop_8900
2025/05/15 19:27:11 INFO Starting managed container gpu=colocated-1-emulated-0 name=live-video-to-video_noop_8910 modelID=noop containerImage=livepeer/ai-runner:live-app-noop
2025/05/15 19:27:11 INFO Watching container container=live-video-to-video_noop_8900
2025/05/15 19:27:16 INFO Container is idle, returning to pool container=live-video-to-video_noop_8900
2025/05/15 19:27:49 INFO Warm container started on loading state, removing from pool on startup container=live-video-to-video_noop_8910
2025/05/15 19:27:49 INFO Watching container container=live-video-to-video_noop_8910
I0515 19:27:49.900391   82080 starter.go:1355] Capability live-video-to-video (ID: Live video to video) advertised with model constraint noop
...
2025/05/15 19:28:04 INFO Container is idle, returning to pool container=live-video-to-video_noop_8910
```

victorges committed:
ai/worker: Wait for LOADING state on container creation

victorges commented:
It works!
```
2025/05/15 14:15:04 INFO Starting managed container gpu=emulated-0 name=live-video-to-video_noop_8900 modelID=noop containerImage=livepeer/ai-runner:live-app-noop
2025/05/15 14:15:08 INFO Health check on container creation returned health="" error="Get \"http://localhost:8900/health\": EOF"
2025/05/15 14:15:10 INFO Health check on container creation returned health="" error="Get \"http://localhost:8900/health\": EOF"
2025/05/15 14:15:12 INFO Health check on container creation returned health="{\"detail\":{\"msg\":\"Failed to retrieve pipeline status.\"}}" error=""
2025/05/15 14:15:14 INFO Health check on container creation returned health="{\"status\":\"LOADING\"}" error=""
2025/05/15 14:15:16 INFO Health check on container creation returned health="{\"status\":\"LOADING\"}" error=""
2025/05/15 14:15:18 INFO Health check on container creation returned health="{\"status\":\"IDLE\"}" error=""
2025/05/15 14:15:18 INFO Watching container container=live-video-to-video_noop_8900
```

victorges committed:
ai/worker: Pick up to date state for IDLE/LOADING checks

victorges committed:
ai/worker: Fix tests

victorges committed:
ai/worker: Reduce flakiness in tests

victorges committed:
ai/worker: Improve O startup time

We actually don't want to wait for LOADING to finish
on the createContainer function as that will block the
O from finishing startup. We should just initialize the
container state instead and let it go IDLE through the
normal logic on watchContainer.
